dkg/src/network.rs

```diff
@@ -0,0 +1,187 @@
+            if let Ok(Ok(Ok(bytes))) = timeout(timeout_duration, rx).await {
+                Ok(protocol.from_bytes(&bytes)?)
+            } else {
+                bail!("self rpc failed");
+            }
+        } else {
+            Ok(self
+                .dkg_network_client
+                .send_rpc(receiver, msg, timeout_duration)
+                .await?)
+        }
+    }
+}
+
+#[async_trait]
+impl RBNetworkSender<DKGMessage> for NetworkSender {
+    async fn send_rb_rpc(
+        &self,
+        receiver: AccountAddress,
+        message: DKGMessage,
+        timeout: Duration,
+    ) -> anyhow::Result<DKGMessage> {
+        self.send_rpc(receiver, message, timeout).await
+    }
+}
+
+pub struct NetworkReceivers {
+    pub rpc_rx: aptos_channel::Receiver<AccountAddress, (AccountAddress, IncomingRpcRequest)>,
+}
+
+pub struct NetworkTask {
+    all_events: Box<dyn Stream<Item = Event<DKGMessage>> + Send + Unpin>,
+    rpc_tx: aptos_channel::Sender<AccountAddress, (AccountAddress, IncomingRpcRequest)>,
+}
+
+impl NetworkTask {
+    /// Establishes the initial connections with the peers and returns the receivers.
+    pub fn new(
+        network_service_events: NetworkServiceEvents<DKGMessage>,
+        self_receiver: aptos_channels::Receiver<Event<DKGMessage>>,
+    ) -> (NetworkTask, NetworkReceivers) {
+        let (rpc_tx, rpc_rx) = aptos_channel::new(QueueStyle::FIFO, 10, None);
+
+        let network_and_events = network_service_events.into_network_and_events();
+        if (network_and_events.values().len() != 1)
+            || !network_and_events.contains_key(&NetworkId::Validator)
+        {
+            panic!("The network has not been setup correctly for DKG!");
+        }
+
+        // Collect all the network events into a single stream
+        let network_events: Vec<_> = network_and_events.into_values().collect();
+        let network_events = select_all(network_events).fuse();
+        let all_events = Box::new(select(network_events, self_receiver));
+
+        (NetworkTask { rpc_tx, all_events }, NetworkReceivers {
+            rpc_rx,
+        })
+    }
+
+    pub async fn start(mut self) {
+        while let Some(message) = self.all_events.next().await {
+            match message {
+                Event::RpcRequest(peer_id, msg, protocol, response_sender) => {
+                    let req = IncomingRpcRequest {
+                        msg,
+                        sender: peer_id,
+                        response_sender: Box::new(RealRpcResponseSender {
+                            inner: Some(response_sender),
+                            protocol,
+                        }),
+                    };
+
+                    if let Err(e) = self.rpc_tx.push(peer_id, (peer_id, req)) {
+                        warn!(error = ?e, "aptos channel closed");
+                    };
+                },
+                _ => {
+                    // Ignored. Currently only RPC is used.
```

Comment:
print error log?

Reply:
if we do anything, it may become a DOS attack vector?